src/Microsoft.Identity.Web/ClientAssertionDescription.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Web
+{
+    /// <summary>
+    /// Description of a client assertion in the application configuration.
+    /// See https://aka.ms/ms-id-web/client-assertions.
+    /// </summary>
+    public class ClientAssertionDescription
+    {
+        /// <summary>
+        /// Constructor of a ClientAssertionDescription.
+        /// </summary>
+        /// <param name="clientAssertionProvider">delegate providing the client assertion
+        /// when it is necessary.</param>
+        public ClientAssertionDescription(Func<CancellationToken, Task<ClientAssertion>> clientAssertionProvider)
+        {
+            ClientAssertionProvider = clientAssertionProvider;
+        }
+
+        /// <summary>
+        /// delegate to get the client assertion.
+        /// </summary>
+        protected Func<CancellationToken, Task<ClientAssertion>> ClientAssertionProvider { get; set; }
+
+        /// <summary>
+        /// Client assertion.
+        /// </summary>
+        private ClientAssertion? _clientAssertion;
+
+        /// <summary>
+        /// Get the signed assertion (and refreshes it if needed).
+        /// </summary>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>The signed assertion.</returns>
+        public async Task<string> GetSignedAssertion(CancellationToken cancellationToken)
+        {
+            if (_clientAssertion == null || (Expiry != null && DateTimeOffset.Now > Expiry))
```

Comment:
I don't think this is going to work in a multi-tenant scenario. We had the same problem in MSAL when we were caching them. 

Scenario: 

1. Get app token for tenant T1 
-> assertion A with `aud`: `l.m.o/T1/oauth/v2.0/token` is generated by invoking the FUNC
2. Get app token for tenant T2
-> assertion A is reused but has bad `aud`

Reply:
This will work in the case for which the feature was provided (cert-less using Managed identity tokens, as the client assertion only depends on the application), but I get the point for other usages.

How did you solve it in MSAL.NET, @bgavrilMS ?
It's not an option to not cache it for MSI, BTW, as MSI will throttle the app after 5 requests for seconds.

Reply:
We disabled assertion caching and rely on token caching only. 

Can you cache the MSI assertion only instead? Is it the same assertion for tokens to all tenants?

`credential.GetTokenAsync(
                    new TokenRequestContext(new[] { "api://AzureADTokenExchange/.default" }, null),
                    cancellationToken);`

Alternatively, let's get this feature into MSAL instead so that you can rely on token caching?